File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.py
import heapq

def solution(jobs):
    job_cnt = len(jobs)  # job 개수
    total_time = 0  # 총 대기 시간 + 작업 시간
    end, i = 0, 0  # 끝나는 시간, jobs 인덱스
    start = -1  # 시작 시간 초기화
    heap = []  # 우선 순위 큐

    jobs.sort()  # 시작 시간 기준으로 정렬
    
    while i < job_cnt or heap:
        # 현재 시간 이전에 시작한 모든 작업들을 힙에 추가
        for j in range(i, job_cnt):
            if start < jobs[j][0] <= end:
                heapq.heappush(heap, (jobs[j][1], jobs[j][0]))
                i += 1
            else:
                break

        # 힙에서 작업을 하나 꺼내 처리
        if heap:
            current_job = heapq.heappop(heap)
            start = end
            end += current_job[0]
            total_time += end - current_job[1]
        else:  # 현재 처리할 작업이 없다면, 다음 작업의 시작 시간으로 점프
            end = jobs[i][0]

    answer = total_time // job_cnt
    return answer

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.js
class MinHeap {
  constructor() {
    this.heap = [null];
  }

  get size() {
    return this.heap.length - 1;
  }

  get min() {
    return this.heap[1];
  }

  swap(idx1, idx2) {
    [this.heap[idx1], this.heap[idx2]] = [this.heap[idx2], this.heap[idx1]];
  }

  heappush(value) {
    this.heap.push(value);
    this.heapifyUp();
  }

  heapifyUp() {
    let idx = this.heap.length - 1;
    while (idx > 1 && this.heap[Math.floor(idx / 2)][0] > this.heap[idx][0]) {
      this.swap(Math.floor(idx / 2), idx);
      idx = Math.floor(idx / 2);
    }
  }

  heappop() {
    if (this.size === 0) {
      return null;
    }
    if (this.size === 1) {
      return this.heap.pop();
    }
    const min = this.heap[1];
    this.heap[1] = this.heap.pop();
    this.heapifyDown(1);
    return min;
  }

  heapifyDown(idx) {
    while (idx * 2 < this.heap.length) {
      let smallerChildIndex = idx * 2;
      if (smallerChildIndex + 1 < this.heap.length && this.heap[smallerChildIndex + 1][0] < this.heap[smallerChildIndex][0]) {
        smallerChildIndex++;
      }

      if (this.heap[idx][0] <= this.heap[smallerChildIndex][0]) {
        break;
      }
      this.swap(idx, smallerChildIndex);
      idx = smallerChildIndex;
    }
  }
}


function solution(jobs) {
  const minHeap = new MinHeap()
  jobs.sort((a, b) => {
    if (a[0] === b[0]) {
      return a[1] - b[1];
    }
    return a[0] - b[0];
  });
  let length = jobs.length
  let totalTime = 0
  let end = 0
  let i = 0

  while (i < length || minHeap.size > 0) {
    // 기간 내에 있는 것 넣기
    // 파이썬처럼 체인 비교가 안된다!!!
    while (i < length && jobs[i][0] <= end) {
      minHeap.heappush([jobs[i][1], jobs[i][0]])
      i++
    }

    // 힙이 비면 다음 작업까지 시간 이동
    if (minHeap.size === 0) {
      if (i < length) {
        end = jobs[i][0]
      }
    } else {
      const [jobTime, jobStart] = minHeap.heappop()
      end += jobTime
      totalTime += end - jobStart
    }
  }

  return Math.floor(totalTime / length)
}

console.log(solution([[0, 3], [1, 9], [2, 6]]))